dango/dex/src/core/market_order.rs

```diff
@@ -0,0 +1,326 @@
+use {
+    crate::{FillingOutcome, LimitOrder, MarketOrder, Order},
+    dango_types::dex::{Direction, OrderId},
+    grug::{
+        IsZero, MultiplyFraction, Number, NumberConst, Signed, StdResult, Udec128, Uint128,
+        Unsigned,
+    },
+    std::{cmp::Ordering, collections::BTreeMap, iter::Peekable},
+};
+
+pub fn match_and_fill_market_orders<M, L>(
+    market_orders: &mut Peekable<M>,
+    limit_orders: &mut Peekable<L>,
+    market_order_direction: Direction,
+    maker_fee_rate: Udec128,
+    taker_fee_rate: Udec128,
+    current_block_height: u64,
+) -> anyhow::Result<Vec<FillingOutcome>>
+where
+    M: Iterator<Item = (OrderId, MarketOrder)>,
+    L: Iterator<Item = StdResult<((Udec128, OrderId), LimitOrder)>>,
+{
+    let mut filling_outcomes = BTreeMap::<OrderId, FillingOutcome>::new();
+
+    // Match the market order to the opposite side of the resting limit order book.
+    let limit_order_direction = -market_order_direction;
+
+    // Find the best offer price in the resting limit order book.
+    // This will be used to compute the market order's worst average execution
+    // price, based on its max slippage.
+    let best_price = match limit_orders.peek_mut() {
+        Some(Ok(((price, _), _))) => *price,
+        Some(Err(e)) => return Err(e.clone().into()),
+        None => return Ok(Vec::new()), // Return early if there are no limit orders
+    };
+
+    // Iterate over the limit orders and market orders until one of them is exhausted.
+    // Since a market order can partially fill a limit order, and that limit order should
+    // remain in the book partially filled, we mutably peek the limit orders iterator and
+    // only advance it when the market order amount is greater than or equal to the remaining
+    // amount of the limit order.
+    //
+    // This is not the case for the market orders. They are matched in the order they were
+    // received, and do not remain after matching is completed.
+    loop {
+        let (price, limit_order_id, limit_order) = match limit_orders.peek_mut() {
+            Some(Ok(((price, limit_order_id), ref mut limit_order))) => {
+                (price, limit_order_id, limit_order)
+            },
+            Some(Err(e)) => return Err(e.clone().into()),
+            None => break,
+        };
+
+        let Some((market_order_id, market_order)) = market_orders.peek_mut() else {
+            break;
+        };
+
+        // Calculate the cutoff price for the current market order
+        let cutoff_price = match market_order_direction {
+            Direction::Bid => Udec128::ONE
+                .checked_add(*market_order.max_slippage)?
+                .checked_mul(best_price)?,
+            Direction::Ask => Udec128::ONE
+                .checked_sub(*market_order.max_slippage)?
+                .checked_mul(best_price)?,
+        };
```

Comment:
@pacmanifold For market BUY orders, this should round **down**. For market SELL orders, this should round **up**. Agree?

Reply:
Nevermind, this is decimal, so there's no rounding...